feat: Add plumbing PR for client side metrics to support the open telemetry instruments

danieljbruce opened the pull request:
**Summary:**

In this PR we introduce all the plumbing that runs the operations from the point where the metrics handler receives data to the point where the data is sent to the Google Cloud Monitoring dashboard.

The diagram below shows all the pieces added in this PR in green:

![2nd plumbing PR](https://github.com/user-attachments/assets/032f145c-589d-41c1-b84d-95f789d42980)

A test stub exists that contains the requests that the metrics collector passes into its metrics handler when a fake method is used. This test stub is used to pass data into the metrics handler in this PR and test the new pipeline that this PR adds.

This is a follow up PR to https://github.com/googleapis/nodejs-bigtable/pull/1566.

**Changes:**

`package.json`: Add the package that contains the MetricServiceClient used for sending data to Google Cloud Monitoring.

`src/client-side-metrics/exporter.ts`: Add the CloudMonitoringExporter which accepts metrics and then converts them to the right format and sends them to the Google Cloud Monitoring dashboard.

`src/client-side-metrics/gcp-metrics-handler.ts`: The metrics handler that stores requests in open telemetry instruments and then periodically sends them to its exporter.

`system-test/cloud-monitoring-exporter.ts`: System tests that write to the google cloud monitoring dashboard with a fixed value passed into the exporter.

`system-test/gcp-metrics-handler.ts`: A system test that passes fixed values into the `gcp-metrics-handler.ts` and ensures it properly writes information to the Google Cloud dashboard.

`test-common/expected-otel-export-input.ts`: Some fixed constants related to values open telemetry provides to the exporter.

`test/metrics-collector/gcp-metrics-handler.ts`: A test that ensures the GCP Metrics handler passes the right information to the exporter.

`test/metrics-collector/metrics-to-request.ts`: A test that ensures the method that converts the value the exporter receives to the value that GCM needs is correct.

**Next Steps:**

- Add client side metrics collection to read-rows operations
- Add application blocking latencies and throttling latencies

danieljbruce committed:
Remove the TODOs

danieljbruce committed:
Add line back

danieljbruce committed:
Add comment

danieljbruce committed:
Add version

danieljbruce committed:
Add version to client side metrics

danieljbruce committed:
linter

danieljbruce committed:
Generate documentation for AttemptInfo interface

danieljbruce committed:
Logger documentation

danieljbruce committed:
Generate more documentation

danieljbruce committed:
Generate documentation

danieljbruce committed:
Make sure test reports correct duration, zone

cluster

danieljbruce committed:
Generate documentation for the dimensions to strin

danieljbruce committed:
Add version to the dimensions

danieljbruce committed:
Fix the client name. The version is going to chan

danieljbruce committed:
Update the expected output file.

danieljbruce committed:
Fox bug, get cluster

danieljbruce committed:
Add fake cluster to tests

danieljbruce committed:
Remove console log

danieljbruce committed:
Generate more documentation

danieljbruce committed:
Require a call to fetch the project when using MT

danieljbruce committed:
use same date provider for all metrics tracers

danieljbruce committed:
In the metrics traceer, don’t fetch the project

for each call

danieljbruce committed:
Remove only

danieljbruce committed:
Merge branch 'main' into 359913994-first-PR-add-infrastructure

danieljbruce committed:
Add open telemetry api

danieljbruce committed:
Merge branch '359913994-first-PR-add-infrastructure' of https://github.com/googleapis/nodejs-bigtable into 359913994-first-PR-add-infrastructure

danieljbruce committed:
Add TestExecuteQuery_EmptyResponse to failures

danieljbruce committed:
TestExecuteQuery_SingleSimpleRow known failures

danieljbruce committed:
Fix syntax in known failures

danieljbruce committed:
Add two tests to the known failures

danieljbruce committed:
TestSampleRowKeys_Retry_WithRetryInfo to known fai

danieljbruce committed:
Change word dimensions to attributes

danieljbruce committed:
Change more docs to use Attributes instead of dim

danieljbruce committed:
attributes

danieljbruce committed:
Test should use attributes as string

danieljbruce committed:
For Windows replace carriage return

danieljbruce committed:
Update documentation with types

danieljbruce committed:
Add metrics collector

danieljbruce committed:
Metrics handler, GCPMetricsHandler and tests add

danieljbruce committed:
Remove only

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-bigtable into 359913994-first-PR-add-infrastructure

# Conflicts:
#	testproxy/known_failures.txt

danieljbruce committed:
Add metrics handlers parameter to Doc

danieljbruce committed:
Don’t require retries to be passed into metrics

collector

danieljbruce committed:
Remove testMeterProvider

danieljbruce committed:
Remove the attributesToString function

danieljbruce committed:
Eliminate unused class

danieljbruce committed:
Generate documentation for the IMetricsHandler

danieljbruce committed:
Generate documentation for GCPMetricsHandler

danieljbruce committed:
Restrict attributes interfaces and solve compile

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

danieljbruce committed:
use undefined instead of null

danieljbruce committed:
Introduce enums for allowable values

danieljbruce committed:
Merge branch '359913994-first-PR-add-infrastructure' of https://github.com/googleapis/nodejs-bigtable into 359913994-first-PR-add-infrastructure

danieljbruce committed:
Add more headers

danieljbruce committed:
Remove only

danieljbruce committed:
Use null to pass values around. Not undefined

danieljbruce committed:
Modify test step

danieljbruce committed:
Add metrics

danieljbruce committed:
Don’t provide first response latency

danieljbruce committed:
Remove firstResponseLatency from operation metrics

danieljbruce committed:
Expose interface allowing undefined not null

danieljbruce committed:
Better explanations for design decision inline

danieljbruce committed:
Use attempt start time not operation start time

for firstResponseLatency

danieljbruce committed:
Adjust tests for first response latency

danieljbruce committed:
Remove TODO

danieljbruce committed:
Use the MethodName enum instead of string

danieljbruce committed:
Don’t use enum for streaming operation

danieljbruce committed:
Remove copy/pasted comment

danieljbruce committed:
Rename to OperationMetricsCollector

danieljbruce committed:
Rename the method to getOperationAttributes

gcf-owl-bot[bot] committed:
Merge branch '359913994-first-PR-add-infrastructure' of https://github.com/googleapis/nodejs-bigtable into 359913994-first-PR-add-infrastructure

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
Merge branch '359913994-first-PR-add-infrastructure' of https://github.com/googleapis/nodejs-bigtable into 359913994-first-PR-add-infrastructure

danieljbruce committed:
Add aggregate views to the GCP metrics handler

danieljbruce committed:
Merge branch '359913994-first-PR-add-infrastructure' of https://github.com/googleapis/nodejs-bigtable into 359913994-first-PR-add-infrastructure

danieljbruce committed:
Adjust test based on enum changes

danieljbruce committed:
Update the documentation to be more descriptive

danieljbruce committed:
Add the state machine to the metrics collector

danieljbruce committed:
Use grpc code to report attempt/operation status

danieljbruce committed:
Remove parameters from JS Documentation

danieljbruce committed:
Update interfaces and some metrics

- Introduce attributes interface for each metric.
- First latency should be recorded per operation not per attempt
- Update tests for firstResponseLatency change

danieljbruce committed:
Documentation for all the different interfaces

danieljbruce committed:
use operation start time as the benchmark

danieljbruce committed:
Final operation status shouldn’t be included per a

danieljbruce committed:
Move OnAttemptCompleteInfo

Remove unused imports too

danieljbruce committed:
Provide AttemptOnlyAttributes in the only file

In the only file that it is needed

danieljbruce committed:
Move over the OperationOnlyAttributes

danieljbruce committed:
Adjust the guard so that it is earlier

danieljbruce committed:
Adjust the test output file

Dates don’t get read for rows after the first row anymore.

danieljbruce committed:
Change streaming back to STREAMING/UNARY

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

danieljbruce committed:
Change metrics handler interface to support each metric

danieljbruce committed:
Revert "Change metrics handler interface to support each metric"

This reverts commit 2781561f3791868855e0daf76d90d4f80736beca.

danieljbruce committed:
Supply the projectId later in the client side

metrics lifecycle

danieljbruce committed:
Remove the GCPMetricsHandler file

danieljbruce committed:
Change location of the client-side-metrics-attribu

file

danieljbruce committed:
Change common test utilities folder name

danieljbruce committed:
Remove aliases for grpc status

danieljbruce committed:
Should be MethodName type

danieljbruce committed:
Rename variable as it expands beyond latency

danieljbruce committed:
Remove private methods for building attributes

danieljbruce committed:
Replace the logger class with a simple object

danieljbruce committed:
Remove only

danieljbruce committed:
Remove the logger classes

Replace them with a simpler object

danieljbruce committed:
Add stubs

danieljbruce committed:
Revert "Remove the GCPMetricsHandler file"

This reverts commit b6f130258101446a7fc65c02a36039184ad9d5ed.

danieljbruce committed:
Add exporter and helper function

danieljbruce committed:
Add headers and fixture

danieljbruce committed:
Add milliseconds unit

danieljbruce committed:
Record to a histogram

danieljbruce committed:
Add necessary elements to the GCPMetricsHandler

# Conflicts:
#	src/client-side-metrics/gcp-metrics-handler.ts

danieljbruce committed:
Pass an exporter into the GCPMetricsHandler

danieljbruce committed:
Move file to tests

danieljbruce committed:
Remove unused import

danieljbruce committed:
Adapt the fixture to include the projectId

danieljbruce committed:
More exporter code fixes

danieljbruce committed:
Add the fixture. We are going to use it later

danieljbruce committed:
Solved compiler errors. Test almost passes

danieljbruce committed:
Address the failing export test

danieljbruce committed:
Fixed the exporter. It now writes data properly.

danieljbruce committed:
Test should complete when the export doesnt error

danieljbruce committed:
Add the fixture to the shared folder

danieljbruce committed:
Remove two files not in this PR

danieljbruce committed:
delete empty file

danieljbruce committed:
Remove files that are already in the other PR

danieljbruce committed:
Remove the metrics handler fixture

danieljbruce committed:
Use 3 metrics instead of 1

danieljbruce committed:
Replace with proper buckets

danieljbruce committed:
Change the metrics handler fixture

danieljbruce committed:
Stop using stub in exporter

danieljbruce committed:
Use more realistic buckets

danieljbruce committed:
Remove the go request to export

danieljbruce committed:
Modify the fixtures to be more realistic

danieljbruce committed:
Change the mean

danieljbruce committed:
Remove only

danieljbruce committed:
Add the export input fixture

danieljbruce committed:
Reducing export interval makes time complete

danieljbruce committed:
Add the GCPMetricsHandler test

danieljbruce committed:
Add a stub for the otel export

danieljbruce committed:
Test calling export is now working

danieljbruce committed:
Remove old unused code

danieljbruce committed:
Change tests and json structure to work

with metricsCollectorData structure

danieljbruce committed:
Update all the fixtures

danieljbruce committed:
Fix the view creation code

Pass in latency buckets

danieljbruce committed:
Starting test for gcp-metrics-handler

danieljbruce committed:
Put tests in the proper places

danieljbruce committed:
Replace start and end time with more recent values

danieljbruce committed:
Utilize the new metrics handler interface

danieljbruce committed:
Solve compiler errors resulting from metrics handl

danieljbruce committed:
Update the fixture

danieljbruce committed:
rewrite the metric to request method

danieljbruce committed:
Add interfaces to work with type guards

danieljbruce committed:
Correct the compile error problems

different metric types are under metrics not scope metrics

danieljbruce committed:
Fix expected OTEL output test

danieljbruce committed:
Remove TODOs

danieljbruce committed:
Fix test to compare pointwise

danieljbruce committed:
connectivity error count corrections

danieljbruce committed:
Correct the fixtures

danieljbruce committed:
Eliminate tests using the old export input fixture

danieljbruce committed:
Add headers

danieljbruce committed:
run linter

danieljbruce committed:
Modify the gcp-metrics-handler and test the proces

end to end

danieljbruce committed:
Remove only

danieljbruce committed:
Use a fake projectId

danieljbruce committed:
Only call export once

danieljbruce committed:
Ensure test suite completes

danieljbruce committed:
Remove shutdown

danieljbruce committed:
remove async

danieljbruce committed:
Don’t export the data twice

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-bigtable into 359913994-exporter-PR

# Conflicts:
#	package.json
#	src/client-side-metrics/client-side-metrics-attributes.ts
#	src/client-side-metrics/metrics-handler.ts
#	test/metrics-collector/typical-method-call.txt

danieljbruce committed:
Increase the timeout

danieljbruce committed:
Use the PushMetricExporter interface

danieljbruce committed:
Removed some interfaces that are not used anymore

danieljbruce committed:
Update JSdoc

danieljbruce committed:
Move fake start time and fake end time

danieljbruce committed:
Remove the TODO

danieljbruce committed:
Update documentation

danieljbruce committed:
Add additional information to the error reported

danieljbruce committed:
Move start time and end time

danieljbruce committed:
Try to use timestamps in order

danieljbruce committed:
Reduce timestamp delay

danieljbruce committed:
Remove only

danieljbruce committed:
Inline addFakeRecentTimestamps

danieljbruce committed:
Move replace timestamps into the only file it

is used

danieljbruce committed:
Fix comment

danieljbruce committed:
Rename the metric types

danieljbruce committed:
Generate documentation for the new classes

danieljbruce committed:
Add documentation for monitored resource

danieljbruce committed:
Generate documentation for the other metrics

danieljbruce committed:
Generate documentation for the constructor

danieljbruce committed:
Get documentation for replaceTimestamps, fixtures

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-bigtable into 359913994-exporter-PR

danieljbruce committed:
Reduce the interval time

danieljbruce committed:
Make view list based on latencies

danieljbruce commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,460 @@
+              type: metricName,
+              labels: metricLabels,
+            },
+            resource: {
```
Note: This is where the monitored resource is provided.

danieljbruce replied:
![image](https://github.com/user-attachments/assets/9c9b396a-e00c-4c47-ab45-826745d504d4)

danieljbruce commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,460 @@
+          );
+          const timeSeries = {
+            metric: {
+              type: metricName,
```
This really ends up being the same thing as in [Python](https://github.com/googleapis/python-bigtable/pull/924/files#diff-142d623e0503afa707461f3bd80d8808a2e10232038c24d7d69ce1df6bd4dbceR180) even though a prefix isn't here. The prefix is just included when we define the open telemetry instruments.

danieljbruce replied:
> I can't remember if you planned to support that or not

Not right now. [there isn't a need to support that right now](https://docs.google.com/document/d/1n_D1RBCYa44QTgtThEYgIQ3vLO0YuGeK8sAsilCwhIE/edit?disco=AAABc_QMooA) so I think we should leave that out to optimize for feature delivery time.

danieljbruce replied:
We don't need to support this at this time as per the discussion [here](https://docs.google.com/document/d/1n_D1RBCYa44QTgtThEYgIQ3vLO0YuGeK8sAsilCwhIE/edit?disco=AAABc_QMooA).

danieljbruce committed:
Add a guard for count

danieljbruce commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+      streamingOperation: data.streamingOperation,
+      clientName: data.clientName,
+    });
+    this.otelMetrics?.retryCount.add(data.retryCount, {
```
Do we only want to log completed attempts if there were retries like in here?

https://github.com/googleapis/python-bigtable/pull/924/files#diff-291971b83bd5a1e863f09438da490a750a39a632869adff64cd1018dac7532c6R181

danieljbruce commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+      finalOperationStatus: data.finalOperationStatus,
+      clientName: data.clientName,
+    });
+    this.otelMetrics?.firstResponseLatencies.record(data.firstResponseLatency, {
```
Should we only do this for ReadRows like in Python?

https://github.com/googleapis/python-bigtable/pull/924/files#diff-291971b83bd5a1e863f09438da490a750a39a632869adff64cd1018dac7532c6R220

danieljbruce commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+      streamingOperation: data.streamingOperation,
+      clientName: data.clientName,
+    });
+    this.otelMetrics?.connectivityErrorCount.add(data.connectivityErrorCount, {
```
In Node we actually calculate the connectivityErrorCount in the Metrics collector, not in the metrics handler.

This is not done the same way in Python FYI:

https://github.com/googleapis/python-bigtable/pull/924/files#diff-291971b83bd5a1e863f09438da490a750a39a632869adff64cd1018dac7532c6R231

daniel-sanche replied:
That seems fine to me. My only concern is this leaves open the possibility of a bug causing it to record connectivity_error_count values > 1, which I don't think should be possible. But I don't think that detail matters too much

I think this field is only supposed to be recorded if it's not 0, but you can double check on that

danieljbruce replied:
Ok. I'll check with Andre.

@andre-sampaio Can you comment on whether connectivityErrorCount should be recorded if it is not 0?

andre-sampaio replied:
I do think it should be recorded even being 0. I'm trying to find some authoritative source for this but I couldn't find any, but this would be consistent with the [java implementation](https://github.com/googleapis/java-bigtable/blob/f9684290bc97ccebf297711aa85e662b5560c4f9/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java#L426-L431). 

An eventual bug that leads the value to be higher than 1 does slightly concern me though, since it could throw us of during an investigation.

danieljbruce replied:
Ok. I just pushed an update to set connectivityErrorCount to 1 instead of incrementing it.

leaf94 commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+          new PeriodicExportingMetricReader({
+            // Export metrics every 10 seconds. 5 seconds is the smallest sample period allowed by
+            // Cloud Monitoring.
+            exportIntervalMillis: 1_000,
```
I assume this is the default (also across all languages). Just double checking : 0

danieljbruce replied:
Thanks for catching this. Based on what I see it actually looks like we should export every 60 seconds. That is what the Python client does:

https://github.com/googleapis/python-bigtable/pull/924/files#diff-142d623e0503afa707461f3bd80d8808a2e10232038c24d7d69ce1df6bd4dbceR105

The comment should be corrected as well.

leaf94 commented on test/metrics-collector/metricsToRequest.ts:
```diff
@@ -0,0 +1,38 @@
+      expectedOtelExportConvertedValue.timeSeries.length
+    );
+    for (let index = 0; index < convertedValue.timeSeries.length; index++) {
+      // We need to compare pointwise because mocha truncates to an 8192 character limit.
```
very helpful comment : )

daniel-sanche commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,462 @@
+        await this.monitoringClient.createTimeSeries(
+          request as ICreateTimeSeriesRequest
+        );
+        // {code: 0} is typically the format the callback expects in the super class.
```
This comment kind of makes me nervous. Are there ever times the callback expects something other than {code: 0}?

danieljbruce replied:
I tested this a while ago and when OpenTelemetry's MetricExporter is successful it provides the callback with this result:

![image](https://github.com/user-attachments/assets/67a4525e-d724-4d50-9775-4493513991ee)

And with this code if the exporter fails then we feed the callback with an error that looks like this:

![image](https://github.com/user-attachments/assets/60f99a55-589a-4c30-a8e2-146b9e69f314)

And when the OpenTelemetry's MetricExporter fails for some reason then the value provided to the callback looks like this:

![image](https://github.com/user-attachments/assets/656e4d5b-8025-467e-a02c-682538239b7d)

I think the point I was trying to make was that the result passed to the callback has a code which in this PR is mainly used to verify that the exporter worked properly.

Now, for this discussion I think it's really important to talk about how the callback gets used. It looks as though the meter provider with the periodic exporter is printing errors to the console when the callback receives errors. This won't crash their program, but it will flood their console with messages (That's how the OTEL metrics provider appears to use the exporter). However, the system tests in the next PR will fail if the exporter is failing to write the metrics to Google Cloud Monitoring (ie. error code not equal to 0).

Do you think it is good enough that we catch this behaviour with a good system test?

danieljbruce replied:
As discussed offline we expect the code to be 0 when the operation sending the metrics to GCP monitoring is successful and non-zero otherwise.

danieljbruce replied:
Comment added

daniel-sanche commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+        resource: new Resources.Resource({
+          'service.name': 'Cloud Bigtable Table',
+          'cloud.provider': 'gcp',
+          'cloud.platform': 'gce_instance',
```
why are you hard-coding `gce_instance`?

danieljbruce replied:
I think this was copy pasted from a Node sample for working with open telemetry.

The main thing here is that this map passed into `Resource` is a ResourceAttributes types which is defined to be just a map of strings to values. It is just used as a mechanism to send persistent data to the exporter which happens to include `projectId`, `instanceId`, `table`, `cluster` and `zone` so for that matter, we can eliminate anything we don't need.

![image](https://github.com/user-attachments/assets/4a2c819a-3517-46df-83e0-eb3ff2b8bd4a)

daniel-sanche commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+          'service.name': 'Cloud Bigtable Table',
+          'cloud.provider': 'gcp',
+          'cloud.platform': 'gce_instance',
+          'cloud.resource_manager.project_id': data.projectId,
```
why do you have both `cloud.resource_manager.project_id` and `monitored_resource.project_id`?

danieljbruce replied:
We can eliminate `'cloud.resource_manager.project_id': data.projectId,` in this case.

daniel-sanche commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+          'monitored_resource.instance_id': data.instanceId,
+          'monitored_resource.table': data.table,
+          'monitored_resource.cluster': data.cluster,
+          'monitored_resource.zone': data.zone,
```
I think you should also be recording `client_name` and `client_uid` here

daniel-sanche replied:
I see. I think I just stored all monitored_resource fields as data_point attributes, and ignored the OTel.Resource layer. I'm fine with this as-is if it works too though (although I do see _syncAttributes is giving you some trouble)

danieljbruce replied:
With some careful casting I got rid of the `_syncAttributes` error and still managed to get rid of that `ExportInput` interface so that problem is gone.

daniel-sanche commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+          }
+        ),
+        applicationBlockingLatencies: meter.createHistogram(
+          'bigtable.googleapis.com/internal/client/application_blocking_latencies',
```
IIRC this should be `application_latencies` for the internal metric

daniel-sanche commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+          }
+        ),
+        clientBlockingLatencies: meter.createHistogram(
+          'bigtable.googleapis.com/internal/client/client_blocking_latencies',
```
I think this should be throttling_latencies

danieljbruce replied:
Done. Note that the metrics fixture doesn't have `throttling_latencies` yet. We intend to add this at a later stage and when we do so any tests are going to flush out these naming errors.

daniel-sanche commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+      instanceId: data.metricsCollectorData.instanceId,
+      table: data.metricsCollectorData.table,
+      cluster: data.metricsCollectorData.cluster,
+      zone: data.metricsCollectorData.zone,
```
Just to be sure: is the system designed so that each Table gets its own instrumentation stack? (That is, is it safe to hard-code the instance and table, since the library can have multiple?)

danieljbruce replied:
> Just to be sure: is the system designed so that each Table gets its own instrumentation stack?

Yes. That's the idea. And admittedly, to make it work this way we have to do something slightly different from the design doc in that each table or perhaps even each [method call](https://github.com/googleapis/nodejs-bigtable/pull/1571/files#diff-b4e03e7f207567c6df97e62d26cb9a8ccb4f410f3901683d7e399ee944c8ba32R349) will get its own metrics handler because there is a one to one relationship between the metric handler and the instruments.

These parameters define the resource which remains the same across the lifecycle of the instrument. Since the resource contains table information then for each table we need a unique resource and therefore a unique instrument set.

In theory, we could allow one metrics handler per client if the metrics handler stored a map of tables to instrument sets instead, but that might make the metrics handler more complicated then it needs to be right now. There is currently no requirement to allow the user to provide a custom metrics handler so I think a metrics handler per table will be fine.

daniel-sanche replied:
We talked about this offline, but my main worry is that multiple tables could cause duplicate exporters/meters, and we could end up duplicating hte data set to the backend. We'll have to make sure there is some kind of isolation between tables here

danieljbruce replied:
Just adding a note here might help. One thing we discussed we wanted to be sure of is that the design doesn't allow for duplicated data. In the current plan, each table will have a gcp metrics handler and each gcp metrics handler will have an exporter.

danieljbruce replied:
The [test run](https://btx.cloud.google.com/invocations/32fcab0c-ab93-4baa-b678-c88052a68fc2/targets/cloud-devrel%2Fclient-libraries%2Fnodejs%2Fpresubmit%2Fgoogleapis%2Fnodejs-bigtable%2Fnode14%2Fsystem-test/log) succeeds, but I want to revisit your idea which involves storing all the dimensions on the OTEL instrument instead of the resource object because I think it's a more loosely coupled design.

danieljbruce replied:
I changed the design so that we can store all dimensions except project on the OTEL attributes. The project still needs to provided per Exporter because it is used in the metric name.

daniel-sanche replied:
Awesome, thanks!

daniel-sanche commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+      clientUid: data.metricsCollectorData.clientUid,
+      finalOperationStatus: data.finalOperationStatus,
+      clientName: data.clientName,
+    });
```
most of these labels are repeated. Can you use spread syntax or something similar to define them in one place?

danieljbruce replied:
Yup. Used spread syntax. Done.

daniel-sanche commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+      clientUid: data.metricsCollectorData.clientUid,
+      finalOperationStatus: data.finalOperationStatus,
+      streamingOperation: data.streamingOperation,
+      clientName: data.clientName,
```
Shouldn't appProfileId, clientUid, and clientName be the same for all operations? If so, can't they be handle in `initialize`?

danieljbruce replied:
> Shouldn't appProfileId, clientUid, and clientName be the same for all operations?

Yes, but in this case I chose to group them with attributes to be consistent with [requests in Golang](https://paste.googleplex.com/5938076009103360). Either way will work and either way has its tradeoffs. What do you think?

danieljbruce replied:
> If so, can't they be handle in initialize

Yes, we could pass them into initialize and store them on the resource in this case.

daniel-sanche commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+      cluster: data.metricsCollectorData.cluster,
+      zone: data.metricsCollectorData.zone,
+    });
+    this.otelMetrics?.attemptLatencies.record(data.attemptLatency, {
```
is there a way to get rid of these optionals? It would be good if we could have some more guarantees here.

Maybe turn `initialize` into `getMetrics`, and have it return the metrics object, creating it if needed?

daniel-sanche replied:
Also, nit: I think Metrics is the wrong terminology for these. I think these are [Instruments](https://opentelemetry.io/docs/specs/otel/metrics/#programming-model)

danieljbruce replied:
> Maybe turn initialize into getMetrics, and have it return the metrics object, creating it if needed?

Yes. I like this. Done.

daniel-sanche commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,313 @@
+      attemptStatus: data.attemptStatus,
+      clientName: data.clientName,
+    });
+    this.otelMetrics?.serverLatencies.record(data.serverLatency, {
```
what happens if serverLatency is null?

danieljbruce replied:
I looked into this and if server latency is null then the metric simply doesn't make it to the export function.

danieljbruce committed:
Invert the for and if

danieljbruce committed:
Pull all attributes out

danieljbruce committed:
Eliminate the need for the type guard

danieljbruce committed:
Eliminate the data points interface

danieljbruce committed:
only

danieljbruce committed:
Revert "Eliminate the need for the type guard"

This reverts commit 791e70dcb71cbf067eac351da6f62ef438be3d5c.

danieljbruce committed:
Pull timeseries into one variable

Solve the descriptor issue

danieljbruce committed:
Eliminate an unused import

danieljbruce commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,462 @@
+  for (const scopeMetrics of exportArgs.scopeMetrics) {
+    for (const metric of scopeMetrics.metrics) {
+      const metricName = metric.descriptor.name;
+      if (isCounterMetric(metric)) {
```
> It looks like there's a lot of duplicate code here

Yes. This could be much better. I pulled the `for` outside of the `if` which allows me to eliminate a lot of the duplicate code. This involved using a new type guard for the metric value so a lot of the code here is much more simplified now.

danieljbruce commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,462 @@
+              ],
+              labels: resourceLabels,
+            },
+            valueType: 'INT64',
```
Note that `INT64` is only used for `retry_count` and `connectivity_error_count`. For all other metrics the 'DISTRIBUTION' value type is used.

danieljbruce replied:
Yeah. I like breaking this stuff down because to the code reader it makes the variable interelationship clear. In particular, helper functions let us know which variables aren't related which makes it simpler to look at.

I added four helpers:
getMetric
getResource
getIntegerPoints
getDistributionPoints

So this function is much shorter now.

danieljbruce committed:
Add a comment that explains the usefulness

of each metric attribute category

danieljbruce commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,462 @@
+            resource: {
+              type: exportArgs.resource._syncAttributes[
+                'monitored_resource.type'
+              ],
```
> Shouldn't it always be bigtable_client_raw?

You're right. In practice it is always `bigtable_client_raw` hardcoded [here](https://github.com/googleapis/nodejs-bigtable/pull/1569/files#diff-0613c4be6e08c6da67824291f9be7e4d391e723f88c1417673066fb91077890aR134) and right now the gcp metrics handler always feeds into this exporter. Technically this is more extensible in case we had a different exporter one day, but it is simpler if we just hardcode it here too.

What do you think?

daniel-sanche replied:
I think I understand what you're going for a bit better now. Personally, since the field is only relevant for the exporter, I think it makes sense to keep it here. Keep all the GCP Cloud Monitoring data in one place, all the OTel configuration in another, etc, instead of mixing them all together

But keeping all the Monitored Resource data together as an OTel Resource[ like you're doing](https://github.com/googleapis/nodejs-bigtable/pull/1569/files#diff-0613c4be6e08c6da67824291f9be7e4d391e723f88c1417673066fb91077890aR134) also makes sense, so I'm fine with it as-is

danieljbruce replied:
I actually ended up going your way on this because I didn't like the design pattern of using the Resource to store export data in retrospect. I think the extra complication is unnecessary.

danieljbruce committed:
Remove the gce instance setting

danieljbruce committed:
Eliminate the gcp cloud provider setting

danieljbruce committed:
run all tests

danieljbruce committed:
Eliminate duplicate project id from monitored

resource

danieljbruce committed:
Change the two metric names

danieljbruce committed:
Extend the timeout so that the exporter has

chance to work

danieljbruce committed:
Use spread syntax

danieljbruce committed:
Changed metric to otel instruments

danieljbruce committed:
Remove optional on Otel instruments

danieljbruce committed:
Rename initialize to getMetrics

danieljbruce commented:
@andre-sampaio For the comment at https://github.com/googleapis/nodejs-bigtable/pull/1569#discussion_r1982385150 could you please provide us with information formally on when we should record the connectivity error count?

danieljbruce commented:
@andre-sampaio @daniel-sanche Tests are failing in CI now due to the Node upgrade, but they are passing locally so this is ready for review again.

danieljbruce commented:
For https://github.com/googleapis/nodejs-bigtable/pull/1569#discussion_r1982351173: I asked the Bigtable team and we don't need to support this yet. Let's prioritize getting the metrics working for ReadRows and we can go from there.

danieljbruce committed:
Pin promisify to version 4

danieljbruce committed:
cast to string - compiler errors

andre-sampaio commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,298 @@
+   */
+  private getMetrics(data: MonitoredResourceData): Metrics {
+    if (!this.otelInstruments) {
+      const latencyBuckets = [
```
have we defined the latency buckets somewhere? I'm trying to find a reference. These are different than the ones we use in the java client: https://github.com/googleapis/java-bigtable/blob/966f46d54d8d7ccd7fd8a5dcb6bb6fcf9ace14e7/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsConstants.java#L67-L73

danieljbruce replied:
Good callout. I modelled this after the Python client [here](https://github.com/googleapis/python-bigtable/pull/924/files#diff-142d623e0503afa707461f3bd80d8808a2e10232038c24d7d69ce1df6bd4dbceR56), but must have missed Mattie's comment below. I'll make this update now.

danieljbruce replied:
Done

danieljbruce commented on src/instance.ts:
```diff
@@ -1310,7 +1310,7 @@ Please use the format 'my-instance' or '${bigtable.projectName}/instances/my-ins
 
     if (policy.etag !== null && policy.etag !== undefined) {
-      (policy.etag as {} as Buffer) = Buffer.from(policy.etag);
+      (policy.etag as {} as Buffer) = Buffer.from(policy.etag as string);
```
Yeah. With the Node 18 upgrade the compiler began complaining so a cast is actually required here.

danieljbruce commented on src/table.ts:
```diff
@@ -1042,7 +1042,7 @@ export class Table extends TabularApiSurface {
 
     if (policy.etag !== null && policy.etag !== undefined) {
-      (policy.etag as {} as Buffer) = Buffer.from(policy.etag);
+      (policy.etag as {} as Buffer) = Buffer.from(policy.etag as string);
```
Yeah. With the Node 18 upgrade the compiler began complaining so a cast is actually required here.

danieljbruce committed:
Change the latency buckets

Adjust the test fixtures as well to account for the latency bucket change.

danieljbruce committed:
Updated the resultCallback comment

daniel-sanche commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,409 @@
+        );
+        const metric = {
+          type: metricName,
+          labels: metricLabels,
```
Maybe I'm misunderstanding node, but is converting this to a new metricLabels object necessary? Shouldn't it be equivalent to datapoint.attributes? (assuming the status/streaming attributes were written in the proper format to begin with)

danieljbruce replied:
In this case it was necessary because the property names upstream were slightly different from the property names here like `client_uid` versus `clientUid`. So I changed the property names upstream so we can just use `datapoint.attributes` because to your point, I think it's nicer if we don't have to do any conversions here.

danieljbruce replied:
After including more attributes in datapoint.attributes like table, cluster and zone this is now a subset of `datapoint.attributes` so I had to add more code here.

daniel-sanche commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,462 @@
+ *
+ * This interface defines the structure of the object that is passed to the
+ * `metricsToRequest` function to convert OpenTelemetry metrics into a format
+ * suitable for the Google Cloud Monitoring API.
```
Can you explain the kinds of errors you're seeing? I see the red in the screenshots, but I'm not sure what the cause is

This is feeling like a bit of a workaround to me. We should try to use the types supplied by OTel if we can. If _syncAttributes is not advertised, maybe it's meant to be internal

danieljbruce replied:
The image below shows the error message I get if I change the type for the argument to `ResourceMetrics` from `ExportInput`. As you can see the compiler says exportArgs.resource doesn't have the property `_syncAttributes`.

<img width="1364" alt="image" src="https://github.com/user-attachments/assets/b2e47f41-3984-4370-8d2d-abfba0145257" />

But when we pass the argument from the `export` function in here then `exportArgs.resource` actually has a `_syncAttributes` property even though the interface doesn't include this property!

<img width="1170" alt="image" src="https://github.com/user-attachments/assets/a08185ae-e9a7-459f-9874-384a10bdadc5" />

Here is another screenshot so you can see exactly what will get passed into export:

<img width="1521" alt="image" src="https://github.com/user-attachments/assets/f4bc7f02-2e0b-4298-8ae9-8cbb3cf3a5d5" />

I could try to make this work an eliminate `ExportInput` by not using the `_syncAttributes` at all, but I don't know what the side effects of that will be. I'll try it and get back to you.

I took another look at the `buckets` error and it looks like we can deal with it by using a cast so I'm not worried about that anymore:
ie. (value as Histogram)

danieljbruce replied:
So I eliminated `ExportInput` and used some careful casting to deal with the `_syncAttributes` error. This eliminates a whole dependency chain of interfaces we don't need anymore.

daniel-sanche commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,409 @@
+    project_id:
+      exportArgs.resource._syncAttributes['monitored_resource.project_id'],
+    table: exportArgs.resource._syncAttributes['monitored_resource.table'],
+    zone: exportArgs.resource._syncAttributes['monitored_resource.zone'],
```
I mentioned above, but seeing `._syncAttributes` seems a little sketchy to me. That doesn't seem to be part of the advertised API surface

danieljbruce replied:
`_syncAttributes` aren't part of the API surface, but they actually exist when the `export` method is called as shown in the screenshot in one of my other comments.

We could be grouping these under the data point attributes instead, but I think in other comments we discussed either way will work?

danieljbruce replied:
I actually ended up deciding to group these under data point attributes mainly because then it doesn't tightly couple the metrics handler to a table.

daniel-sanche commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,409 @@
+        // to 0 then the operation failed. Open telemetry logs errors to the
+        // console when the resultCallback passes in non-zero code values and
+        // logs nothing when the code is 0.
+        const exportResult = {code: 0};
```
can you move this constant to the top of the file?

danieljbruce replied:
It's only used in one place. I just eliminated it and replaced it inline on the next line of code.

daniel-sanche commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,409 @@
+                },
+              ],
+              unit: scopeMetric.descriptor.unit || 'ms', // Default to 'ms' if no unit is specified
+            };
```
Can this be turned into an if? I think long ternary statements like this can be hard to follow

danieljbruce replied:
I used ternary here because it's more _functional_, but I see your point on readability. It's back to an if statement now.

daniel-sanche commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,409 @@
+              (allAttributes as OnAttemptAttribute).attemptStatus?.toString() ??
+              (
+                allAttributes as OnOperationAttribute
+              ).finalOperationStatus?.toString(),
```
instead of doing toString here, can't you write it as a string to start with?

danieljbruce replied:
Yes. I pushed the string conversion back to the metrics collector [here](https://github.com/googleapis/nodejs-bigtable/pull/1569/commits/a587d0c97b83b061b20016cf5edb062049284d10) and use status now instead of attemptStatus and finalOperationStatus.

danieljbruce committed:
Change the test description

danieljbruce committed:
Add a comment explaining what the test is about

danieljbruce committed:
Completed the test for two metrics handlers

danieljbruce committed:
remove only

danieljbruce committed:
Cast as histogram to make compile error go away

danieljbruce committed:
Move the duplicate copies of replaceTimestamps

into one file

danieljbruce committed:
Take steps to eliminate the ExportInput interface

danieljbruce committed:
Add a header

danieljbruce committed:
Remove only

danieljbruce commented:
For threads like [this](https://github.com/googleapis/nodejs-bigtable/pull/1569#discussion_r1994371184) it is not letting me comment in the conversation tab

danieljbruce commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,462 @@
+  finalOperationStatus: number;
+  streamingOperation?: string;
+  clientName: string;
+}
```
By eliminating `ExportInput` I was able to eliminate the need for `OnOperationAttribute` and `OnAttemptAttribute` along with a number of interfaces the `ExportInput` was counting on.

danieljbruce committed:
Eliminate ExportInput and dependencies

danieljbruce committed:
Eliminate constant

danieljbruce committed:
Push conversion to status back to metrics collect

Also add if statement for timeseries.

danieljbruce committed:
Change property names to match MetricServiceClient

expectation

danieljbruce committed:
Solve the app_profile_id issue

danieljbruce committed:
It is actually app_profile not app_profile_id

danieljbruce committed:
Add guards that stop the export call

danieljbruce committed:
Remove only

danieljbruce committed:
Add a new test for ten metrics handlers

danieljbruce committed:
Remove only

danieljbruce committed:
Do not pass data through the Resource object

anymore

danieljbruce committed:
Add a test for writing duplicate points to MH

danieljbruce committed:
Eliminate interface

danieljbruce committed:
Set connectivity error count to 1 (not increment)

daniel-sanche reviewed (commented):
This is looking very solid! My only remaining request would be to rename the Metrics interface to `Instruments` or `MetricInstruments`, since [Metrics has a different meaning in OTel](https://opentelemetry.io/docs/concepts/signals/metrics/)

After that, this should be good to merge

danieljbruce committed:
Latency improvement

serverTime and connectivityErrorCount should only be read once.

danieljbruce committed:
Add latency optimizations and update fixtures

danieljbruce committed:
Update fixtures with new latency measurements

The optimization means some code gets skipped therefore lower latency in the tests

danieljbruce committed:
Check for first response latency first

danieljbruce commented on src/client-side-metrics/exporter.ts:
```diff
@@ -0,0 +1,250 @@
+                },
+              },
+            ],
+          });
```
The tests that export results to Google Cloud Monitoring succeed at exporting the metrics so I don't think so. If we did have a unit for `connectivity_error_count` and `retry_count` then what would it be? These are just counters.

danieljbruce committed:
Break metricsToRequest down into smaller fns

danieljbruce committed:
Refactor the metricToRequest function

danieljbruce committed:
Rename interface

danieljbruce commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,272 @@
+ * A collection of OpenTelemetry metric instruments used to record
+ * Bigtable client-side metrics.
+ */
+interface Metrics {
```
Done

danieljbruce committed:
Rename to getInstruments

danieljbruce commented on src/client-side-metrics/gcp-metrics-handler.ts:
```diff
@@ -0,0 +1,272 @@
+   * which will be provided to the exporter in every export call.
+   *
+   */
+  private getMetrics(projectId: string): Metrics {
```
Yup. Renamed to getInstruments.

danieljbruce commented:
Re: https://github.com/googleapis/nodejs-bigtable/pull/1569#discussion_r2006465172. I think this is the MetricsExporter file? We can change `{code: 0}` elsewhere in `src` if you see it in another place, but I can't find one. `{code: 0}` appears in a lot of tests though.

daniel-sanche reviewed (approved):
LGTM!
